File: src/pipeline.py
import asyncio
import logging
import uuid
from datetime import datetime, timezone

from src.config import Config
from src.db import Database
from src.fetcher import Fetcher
from src.filters import Filters
from src.generator import ReplyGenerator
from src.poster import Poster
from src.telegram_bot import TelegramReviewBot
from src.tweeter import Tweeter

logger = logging.getLogger(__name__)


class Pipeline:
    def __init__(self, db: Database, fetcher: Fetcher, filters: Filters, generator: ReplyGenerator, poster: Poster, telegram: TelegramReviewBot, tweeter: Tweeter = None):
        self.db = db
        self.fetcher = fetcher
        self.filters = filters
        self.generator = generator
        self.poster = poster
        self.telegram = telegram
        self.tweeter = tweeter
        self.loop = asyncio.new_event_loop()

    def run(self) -> dict:
        stats = {"fetched": 0, "drafts_created": 0, "skipped": 0}

        reply_count = self.db.get_today_reply_count()
        if reply_count >= Config.DAILY_REPLY_CAP:
            logger.info("Daily cap reached (%d). Skipping.", Config.DAILY_REPLY_CAP)
            stats["skipped_reason"] = "daily_cap_reached"
            return stats

        # Collect tweets
        tweets = []
        for keyword in self.db.get_active_keywords():
            tweets.extend(self.fetcher.search_keyword(keyword))
        for account in self.db.get_active_accounts():
            tweets.extend(self.fetcher.fetch_account_tweets(account))

        # Dedup
        tweets = self.filters.dedup(tweets)
        stats["fetched"] = len(tweets)
        logger.info("Fetched %d new tweets", len(tweets))

        # Filter and generate drafts
        for tweet in tweets:
            if reply_count + stats["drafts_created"] >= Config.DAILY_REPLY_CAP:
                break

            if not self.filters.filter_tweet(tweet):
                stats["skipped"] += 1
                continue

            reply_text = self.generator.generate(
                tweet_content=tweet["content"],
                author_username=tweet["author_username"],
                author_bio=tweet.get("author_bio", ""),
                thread_context=tweet.get("thread_context"),
            )

            if reply_text is None:
                stats["skipped"] += 1
                continue

            # Save tweet
            tweet_to_save = {k: v for k, v in tweet.items() if k != "metrics"}
            self.db.save_tweet(tweet_to_save)

            # Save reply as pending (NOT posting yet)
            reply_id = str(uuid.uuid4())
            self.db.save_reply({
                "id": reply_id,
                "tweet_id": tweet["tweet_id"],
                "draft_text": reply_text,
                "status": "pending",
                "created_at": datetime.now(timezone.utc).isoformat(),
            })

            # Send to Telegram for approval
            self.loop.run_until_complete(self.telegram.send_review(tweet, reply_text, reply_id))
            stats["drafts_created"] += 1
            logger.info("Draft created for tweet %s", tweet["tweet_id"])

        # Run tweeter
        stats["tweeter"] = self.run_tweeter()
        logger.info("Pipeline complete: %s", stats)
        return stats

    def run_tweeter(self) -> dict:
        if not self.tweeter:
            return {"skipped_reason": "tweeter_not_configured"}
        return self.tweeter.run()

File: src/pipeline.ts
import { v4 as uuidv4 } from 'uuid';
import { db } from './db';
import { fetcher } from './fetcher';
import { filters } from './filters';
import { generator } from './generator';
import { poster } from './poster';
import { telegram } from './telegram';
import { relevanceScorer } from './relevance-scorer';
import { Tweeter, StatsCollector } from './tweeter';
import { config } from './config';

interface PipelineStats {
  fetched: number;
  drafts_created: number;
  skipped: number;
  skipped_reason?: string;
  tweeter?: any;
}

export class Pipeline {
  private tweeter: Tweeter;

  constructor() {
    const statsCollector = new StatsCollector(config.agentlynx_api_url);
    this.tweeter = new Tweeter(statsCollector);
  }

  async run(): Promise<PipelineStats> {
    const stats: PipelineStats = {
      fetched: 0,
      drafts_created: 0,
      skipped: 0,
    };

    const replyCount = await db.getTodayReplyCount();
    if (replyCount >= config.DAILY_REPLY_CAP) {
      console.log(`[INFO] Daily cap reached (${config.DAILY_REPLY_CAP}). Skipping.`);
      stats.skipped_reason = 'daily_cap_reached';
      return stats;
    }

    // Collect tweets
    const tweets = [];
    const keywords = await db.getActiveKeywords();
    for (const keyword of keywords) {
      tweets.push(...(await fetcher.searchKeyword(keyword)));
    }

    const accounts = await db.getActiveAccounts();
    for (const account of accounts) {
      tweets.push(...(await fetcher.fetchAccountTweets(account)));
    }

    // Dedup
    const dedupTweets = filters.dedup(tweets);
    stats.fetched = dedupTweets.length;
    console.log(`[INFO] Fetched ${dedupTweets.length} new tweets`);

    if (dedupTweets.length === 0) {
      console.log('[INFO] Pipeline complete: no new tweets');
      return stats;
    }

    // Filter tweets: only last 24 hours
    const now = new Date();
    const twentyFourHoursAgo = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    const recentTweets = dedupTweets.filter((tweet) => {
      const tweetTime = new Date(tweet.created_at);
      return tweetTime >= twentyFourHoursAgo;
    });

    console.log(`[INFO] Filtered to ${recentTweets.length} tweets from last 24 hours`);

    if (recentTweets.length === 0) {
      console.log('[INFO] No tweets from last 24 hours');
      stats.skipped = dedupTweets.length;
      return stats;
    }

    // Apply additional filters
    const filteredTweets = [];
    for (const tweet of recentTweets) {
      if (await filters.filterTweet(tweet)) {
        filteredTweets.push(tweet);
      } else {
        stats.skipped++;
      }
    }

    console.log(`[INFO] Filtered down to ${filteredTweets.length} tweets`);

    if (filteredTweets.length === 0) {
      console.log('[INFO] No tweets passed filtering');
      return stats;
    }

    // Find the single best tweet by relevance score
    const bestTweet = await relevanceScorer.findBestTweet(filteredTweets);
    if (!bestTweet) {
      console.log('[INFO] Could not determine best tweet');
      return stats;
    }

    // Generate and post reply for the best tweet only
    console.log(`[INFO] Generating reply for best tweet ${bestTweet.tweet_id}`);
    const replyText = await generator.generate(
      bestTweet.content,
      bestTweet.author_username,
      bestTweet.author_bio,
      bestTweet.thread_context
    );
    console.log(`[INFO] Reply generated: ${replyText ? replyText.substring(0, 50) : 'None'}`);

    if (!replyText) {
      stats.skipped++;
      return stats;
    }

    // Save tweet
    await db.saveTweet(bestTweet);

    // Post reply immediately
    const replyId = uuidv4();
    const success = poster.postReply(bestTweet.tweet_id, replyText);

    // Save reply with status
    const status: 'posted' | 'failed' = success ? 'posted' : 'failed';
    await db.saveReply({
      id: replyId,
      tweet_id: bestTweet.tweet_id,
      draft_text: replyText,
      final_text: null,
      status,
      reviewed_at: null,
      posted_at: success ? new Date().toISOString() : null,
      created_at: new Date().toISOString(),
    });

    // Send result to Telegram
    const label = success ? '✓ Posted' : '✗ Failed';
    await telegram.sendResult(bestTweet, replyText, label);
    stats.drafts_created++;
    console.log(`[INFO] Reply ${label} for tweet ${bestTweet.tweet_id}`);

    // Always try to tweet one original tweet (A, B, or C)
    console.log('[INFO] Running tweeter for original tweet...');
    try {
      const tweeterResult = await this.tweeter.run();
      stats.tweeter = tweeterResult;
    } catch (err) {
      console.error('[ERROR] Tweeter failed:', err);
    }

    console.log(`[INFO] Pipeline complete: ${JSON.stringify(stats)}`);
    return stats;
  }
}

export const pipeline = new Pipeline();
